Tidying up and tweaking mplot3d examples [MEP12]

TrishGillett opened the pull request:
This PR is the follow up to #6303 and covers clean up work for all the mplot3d examples not covered in that one.  It also includes touch ups to an example cleaned up previously, 2dcollections3d_demo.  A lot of my changes are docstring and comment additions, but I've also done some rearranging and refactoring in hopes of improving clarity.  I'll make line comments on some of the points of interest.

discardthree commented on examples/mplot3d/rotate_axes3d_demo.py:
```diff
@@ -1,13 +1,21 @@
 for angle in range(0, 360):
     ax.view_init(30, angle)
     plt.draw()
+    plt.pause(.001)
```
On my computer, this line was needed to make the example viewable.  I got the idea for it when I did the wire3d_animation_demo example, but I'm not sure this is the correct practice since in both examples this line seems to trigger a deprecation warning: 
```
/Users/trish/anaconda/lib/python2.7/site-packages/matplotlib-2.0.0b1+1720.ga8ad349.dirty-py2.7-macosx-10.5-x86_64.egg/matplotlib/backend_bases.py:2444: MatplotlibDeprecationWarning: Using default event loop until function specific to this GUI is implemented
  warnings.warn(str, mplDeprecation)
```
Let me know if there's something else I should do with this instead.

WeatherGod replied:
Which backend are you using?

discardthree replied:
Dumb question: how do I tell?

WeatherGod replied:
`import matplotlib; print(matplotlib.get_backend())`

discardthree replied:
Thanks!  Qt4Agg.

discardthree replied:
If I'm going to make an issue for this, which part is the issue?  That nothing displays for me without the use of plt.pause(), or that the use of plt.pause() triggers that warning?  I'm not really clear on whether plt.draw() is supposed to be sufficient in an example like this.

QuLogic replied:
This may be because `plt.draw` is now lazy and won't draw until idle?

tacaswell replied:
@QuLogic You are correct.  `plt.pause()` spins the GUI event loop so the update gets processed.

I don't think we need to worry about the warning in this PR, it is a know issue.

discardthree commented on examples/mplot3d/tricontourf3d_demo.py:
```diff
@@ -30,7 +34,11 @@
+ax.tricontourf(triang, z, cmap=plt.cm.CMRmap)
+
+# Customize the view angle so it's easier to understand the plot.
+ax.view_init(elev=45.)
```
Same color map and elevation tweaks as in tricontour3d example.

discardthree commented on examples/mplot3d/trisurf3d_demo2.py:
```diff
@@ -1,11 +1,28 @@
-u = (np.linspace(0, 2.0 * np.pi, endpoint=True, num=50) * np.ones((10, 1))).flatten()
-v = np.repeat(np.linspace(-0.5, 0.5, endpoint=True, num=10), repeats=50).flatten()
+
+fig = plt.figure(figsize=plt.figaspect(0.5))
```
I made the two plots in this example display side by side as subplots.  Previously the code was only actually displaying one of them when run.

WeatherGod commented on examples/mplot3d/wire3d_animation_demo.py:
```diff
@@ -1,38 +1,45 @@
-Z = generate(X, Y, 0.0)
 
+# Set the z axis limits so they aren't recalculated each frame.
+ax.set_zlim(-1, 1)
```
That shouldn't happen. Makes me wonder if it is a bug with the style changes for 2.0.

discardthree replied:
I'm doing a git bisect to try and nail down when this happened.

discardthree replied:
Using this example as a test script, I see the same behaviour with versions 1.5.2, 1.5.1, 1.5.0, 1.4.0, 1.2.0, and 1.0.0.  I haven't found a commit where this behaviour doesn't happen.

WeatherGod replied:
ok, how about this issue and the other one with regards to "plt.pause()"
get filed as issues? Let's take out those two changes for now and keep it
focused on just MEP12. We'll have to investigate those two issues further
to see if there is something deeper going on.

On Tue, Jul 5, 2016 at 11:14 AM, Trish Gillett-Kawamoto <
notifications@github.com> wrote:

> In examples/mplot3d/wire3d_animation_demo.py
> <https://github.com/matplotlib/matplotlib/pull/6690#discussion_r69580992>:
>
> >
> > +# Set the z axis limits so they aren't recalculated each frame.
> > +ax.set_zlim(-1, 1)
>
> Using this example as a test script, I see the same behaviour with
> versions 1.5.2, 1.5.1, 1.5.0, 1.4.0, 1.2.0, and 1.0.0. I haven't found a
> commit where this behaviour doesn't happen.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/6690/files/6a5def2d4d237346f47e4e9da592f05058b1a9f9#r69580992>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AARy-L05A6cqDxNM8ebAlLR7HrjuixW9ks5qSnTcgaJpZM4JE0ZQ>
> .
>

discardthree replied:
I'm just realizing I was thinking of the rotate_axes3d example when I wrote the note saying the z limits changed 'as the rotation progressed'.  Sorry for that confusion.  In fact it's the data that's changing in this example, not the rotation.  As the largest z value contracts from say 1 to 0.6, the z limits are self-adjusting so that the extremes of the surface are always using the full z range.  In that updated context, would you say that the behaviour is as expected or does something still sound fishy?

tacaswell replied:
@TrishGillett That behavior is expected given how this code works.

tacaswell commented:
Restarted the failing tests (which are almost certainly un-related).

NelleV commented:
I have also restarted the failing test as it seemed unrelated.

Thanks for the patch @TrishGillett ! The examples are much better.
If you ever work on the examples again, do you mind having the following structure of docstrings:

```python
"""
title

description
"""
```

It'll make our life easier when we move to sphinx-gallery, as this is the required format for examples docstring and sphinx-gallery.

TrishGillett commented:
Thanks for the approvals!  Duuly noted about the docstring structure, I'll check it too from now on. :)

NelleV commented:
The travis failure is unrelated, but I have no clue what's going on: it seems a font file is missing.
@tacaswell any idea?

TrishGillett commented:
Any chance a rebase might help?

tacaswell commented:
A rebase might help.  Not sure what is going wrong with that 2.7 test....

TrishGillett commented:
I'll do a rebase tonight or tomorrow.  While I'm at it, I can change the docstrings like @NelleV mentioned.  Is the idea that the 'title' should match the filename (ex. 2dcollections3d_demo.py => '2D Collections 3D Demo')?  If not, can you give me an example of what you're going for?

NelleV commented:
First, don't feel obliged to do this. Your work on this PR is already a huge improvements of our examples.

In short, a blocker for one of my pull request is the migration of our gallery to sphinx-gallery. Unfortunately, sphinx gallery requires that docstring be of the format "title + description". The title is used both in the gallery as a title to each vignette, and as the title of the docstring rendered for each example.
Here are some examples of titles: http://sphinx-gallery.readthedocs.io/en/latest/

I'll go other some of the plots to make suggestions.

NelleV commented on examples/mplot3d/surface3d_demo.py:
```diff
@@ -1,23 +1,38 @@
+'''
```
A title on that PR could be "3D surfaces"

TrishGillett commented:
It's no problem to add the titles, my goal when I do MEP12 PRs is to try and make the examples 'compliant' with all the standards we're shooting for, so I'd rather do it as part of that process than leave it for someone else to come back and do later.  Maybe you should modify the 'Add introductory sentence or paragraph in the main docstring' point in the guidelines [here](http://matplotlib.org/devel/MEP/MEP12.html) to reflect what you need so some MEP12ers can help you chip away at the problem. :)

NelleV commented:
That's on my todo list :)

TrishGillett commented:
Added example titles and rebased on master.  Travis is running clean now but Appveyor has some complaints.  Can someone in the know check if they're related or not?

efiring commented:
Appveyor failures were unrelated.

efiring commented:
backported to v2.x as a869cc0

efiring commented:
Thank you, @TrishGillett.
